public/app/plugins/datasource/loki/queryUtils.ts

```diff
@@ -295,3 +319,113 @@ export function getStreamSelectorsFromQuery(query: string): string[] {
 
   return labelMatchers;
 }
+
+const PARTITION_LIMIT = 60;
+export function partitionTimeRange(range: TimeRange, timeShift = 1, unit: DurationUnit = 'm'): TimeRange[] {
+  const partition: TimeRange[] = [];
+  for (let from = dateTime(range.from), to = dateTime(from).add(timeShift, unit); to <= range.to && from < range.to; ) {
+    partition.push({
+      from,
+      to,
+      raw: { from, to },
+    });
+
+    /**
+     * The user can request for any arbitrary time range. If we receive one from too
+     * long ago, this will cause an extremely large loop which could break the app.
+     * Additionally, data retention is limited, so ranges that go above PARTIION_LIMIT
+     * will be ignored.
+     */
+    if (partition.length > PARTITION_LIMIT) {
+      return [range];
+    }
+
+    from = dateTime(to);
+    to = dateTime(to).add(timeShift, unit);
+
+    if (to > range.to) {
+      to = range.to;
+    }
+  }
+
+  return partition;
+}
+
+export function requestSupportsPartitioning(queries: LokiQuery[]) {
+  /*
+   * For now, we would not split when more than 1 query is requested.
+   */
+  if (queries.length > 1) {
+    return false;
+  }
+
+  /**
+   * Disable logs volume queries.
+   */
+  if (queries[0].refId.includes('log-volume-')) {
+    return false;
+  }
+
+  if (isLogsQuery(queries[0].expr)) {
+    return false;
+  }
+
+  if (isRateQuery(queries[0].expr)) {
+    return false;
+  }
+
+  return true;
+}
+
+export function combineResponses(currentResult: DataQueryResponse | null, newResult: DataQueryResponse) {
+  if (!currentResult) {
+    return newResult;
+  }
+
+  newResult.data.forEach((newFrame) => {
+    const currentFrame = currentResult.data.find((frame) => frame.name === newFrame.name);
+    if (!currentFrame) {
+      currentResult.data.push(newFrame);
+      return;
+    }
+    combineFrames(currentFrame, newFrame);
+  });
+
+  return currentResult;
+}
+
+function combineFrames(dest: DataQueryResponseData, source: DataQueryResponseData) {
+  const totalFields = dest.fields.length;
+  for (let i = 0; i < totalFields; i++) {
+    dest.fields[i].values.buffer = [].concat.apply([], [source.fields[i].values.buffer, dest.fields[i].values.buffer]);
+  }
+  dest.length += source.length;
+  if (dest.meta.stats) {
+    dest.meta.stats.forEach((stat: QueryResultMetaStat, i: number) => {
+      // We assume same order and meaning of meta stats.
```

Comment:
i'm sorry but i don't think we can assume same order always 😿 .

honestly, i wonder if we should just delete the stats in the chunking-situation for now.
the reason is, i think we cannot even add these together always.

for example, one of the stats is "bytes processed per second". that cannot be just added together (we either need to average them, or max-value or something),

so maybe just delete it for now?

Reply:
I can't answer that. I'll add it for discussion in our call today.

Reply:
Now we combine them without assuming the same order. A follow up to this PR should include a smarter way of dealing with stats and the query inspector.